### PR TITLE
Wip gc get source and dest zones

### PR DESCRIPTION
**Pull request by AlvinYangZF**
Merge @shampoo365 ‘s commit into this branch and commit to master together
@soumendus this pull requset fixed some of the issues that your comments mentioned.

### CODE REVIEW

File: fs/io_zenfs.cc [hunk 1]
```diff
@@ -659,6 +660,64 @@ size_t ZonedRandomAccessFile::GetUniqueId(char* id, size_t max_size) const {
...
+  std::vector<Zone*>::iterator zone_it;
+  for (zone_it = merge_zone_list.begin(); zone_it != merge_zone_list.end();
+       zone_it++) {
+    Zone* zone_idx;
+    zone_idx = *zone_it;
+
+    IOStatus s;
+    s = zone_idx->Reset();
+    if (!s.ok()) {
+      // Debug(logger_, "Failed resetting zone when executing GC!");
```
Comment (skyzh) [inline fs/io_zenfs.cc:700]: Why comment this out? I believe it is needed to report an error if reset failed.

File: fs/io_zenfs.cc [hunk 2]
```diff
@@ -659,6 +660,64 @@ size_t ZonedRandomAccessFile::GetUniqueId(char* id, size_t max_size) const {
...
+
+std::vector<Zone*> ZenFSGCWorker::MarkZonesToMergeData() {
+  merge_zone_list = zbd_->GetReclaimZones();
+  return merge_zone_list;
+}
+
+std::vector<Zone*> ZenFSGCWorker::GetDestZoneToMoveValidData(
+    uint64_t ttl_residue) {
+  while (ttl_residue) {
+    Zone* zone = zbd_->AllocateZone(Env::WLTH_NOT_SET);
```
Comment (skyzh) [inline fs/io_zenfs.cc:713]: Do we have any consideration on zone lifetime here? If leave this with not set, new extents with different lifetime will be pushed onto this zone, making it (possibly) harder to collect garbage.

Comment (skyzh) [reply fs/io_zenfs.cc:713]: And what if zone allocation fails?

File: fs/zbd_zenfs.cc [hunk 3]
```diff
@@ -174,6 +174,19 @@ Zone *ZonedBlockDevice::GetIOZone(uint64_t offset) {
   return nullptr;
 }
 
+std::vector<Zone *> ZonedBlockDevice::GetReclaimZones(void) {
+  std::vector<Zone *> GCReclaimZones;
+  for (const auto z : io_zones) {
+    if ((!z->IsUsed()) && (!z->IsFull())) continue;
+    /* valid data in a zone is larger than 10% max capacity */
+    if ((z->max_capacity_ / z->used_capacity_ >= 10)) {
```
[the remark this answers is not in the document]
Comment (AlvinYangZF) [reply fs/zbd_zenfs.cc:182]: The 10% valid data in a zone should be described as z->used_capacity_ / z->max_capacity, but I didn't want to use floating varables, so I exchanged them. However, it is not readable and the denominator could be 0, so I think it is better to correct it.

[the remark this answers is not in the document]
Comment (skyzh) [reply fs/zbd_zenfs.cc:182]: You can use something like `0.1 * used_capacity_`, and I believe the compiler is smart enough to optimize it into integer multiplication.

File: fs/zbd_zenfs.cc [hunk 4]
```diff
@@ -174,6 +174,20 @@ Zone *ZonedBlockDevice::GetIOZone(uint64_t offset) {
   return nullptr;
 }
 
+std::vector<Zone *> ZonedBlockDevice::GetReclaimZones(void) {
+  std::vector<Zone *> GCReclaimZones;
+  for (const auto z : io_zones) {
+    if ((!z->IsUsed()) && (!z->IsFull())) continue;
+    /* once the valid data in a zone is less than 10%, it will be seleted as GC
+     * zone */
+    if ((z->used_capacity_ * ZENFS_GC_THREASHOLD) / z->max_capacity_ <= 1) {
```
Comment (skyzh) [inline fs/zbd_zenfs.cc:183]: From my perspective, GC is not called very frequently, and we might need to dynamically configure the threshold. So I would prefer to simply use double multiplication.

**Review by skyzh — COMMENTED** [review]
Rest LGTM. There might be some concurrency issues inside, but the current codebase is not complete enough to conduct a full review. I will find and review these issues later.

File: fs/io_zenfs.cc (hunk 1 above)
Comment (AlvinYangZF) [reply fs/io_zenfs.cc]: Added it back

**Review by skyzh — APPROVED** [review]
LGTM for this part. Thanks for your work.

**Pull request merged**